api/server/v1/error.go

```diff
@@ -17,98 +17,292 @@ package api
+//      * Client code calls `FromStatusError` to reconstruct TigrisError from GRPC status and it's payloads
+//        (Extended code is taken from errdetails.ErrorInfo and retry delay from errdetails.RetryInfo)
+//
+// 2. HTTP interface
+// So as HTTP interface is intended to be inspected by users, we marshal HTTP errors
+// in a more human-readable form, less verbose then GRPC errors.
+//
+// Example HTTP error:
+// {
+//   "error": {
+//      "code": "ALREADY_EXISTS"
+//      "message": "database already exists"
+//      "retry_delay": "1s"
+//   }
+// }
+//
+// The flow:
+//		* Server uses `api.Errorf({tigris code}, ...)` to report a TigrisError
+//      * We provide TigrisError.As(*runtime.HTTPStatusError) to be able to override HTTP
+//        status code in GRPC gateway if need to
+//      * The error is converted to GRPC status by GRPCStatus() interface
+//      * We setup (see CustomMarshaller) and provide custom marshaller of
+//        the GRPC status (see MarshalStatus) in GRPC gateway
+//      * Client parses the error using UnmarshalStatus into TigrisError
+
 // TigrisError is our Tigris HTTP counterpart of grpc status. All the APIs should use this Error to return as a user facing
 // error. TigrisError will return grpcStatus to the muxer so that grpc client can see grpcStatus as the final output. For
 // HTTP clients see the **MarshalStatus** method where we are returning the response by not the grpc code as that is not
 // needed for HTTP clients.
 type TigrisError struct {
 	// The status code, which should be an enum value of [google.rpc.Code][google.rpc.Code]. We don't need to marshal
 	// this code for HTTP clients.
-	Code codes.Code `json:"code,omitempty"`
+	Code Code `json:"code,omitempty"`
 	// A developer-facing error message.
 	Message string `json:"message,omitempty"`
-	// A list of messages that carry the error details. This is mainly to send our internal codes and messages.
-	Details []*ErrorDetails `json:"details,omitempty"`
+
+	Details []proto.Message `json:"details,omitempty"`
 }
 
 // Error to return the underlying error message
 func (e *TigrisError) Error() string {
 	return e.Message
 }
 
-// GRPCStatus converts the TigrisError and return status.Status. This is used to return grpc status to the grpc clients
-func (e *TigrisError) GRPCStatus() *status.Status {
-	s := &spb.Status{
-		Code:    int32(e.Code),
-		Message: e.Message,
-	}
+// WithDetails a helper method for adding details to the TigrisError
+func (e *TigrisError) WithDetails(details ...proto.Message) *TigrisError {
+	e.Details = append(e.Details, details...)
+	return e
+}
 
-	if len(e.Details) == 0 {
-		return status.FromProto(s)
+// WithRetry attached retry information to the error
+func (e *TigrisError) WithRetry(d time.Duration) *TigrisError {
+	if d != 0 {
+		e.Details = append(e.Details, &errdetails.RetryInfo{RetryDelay: durationpb.New(d)})
```

Comment:
I thought retry info is outside?

Reply:
Don't get it. Details contains any extended information we want to carry around.